File: Snippet3/Snippet3/guides_ios_managing-push-notification_push-to-user_sending-messages.m
//
//  guides_ios_managing-push-notification_push-to-user_sending-messages.m
//  Snippet3
//
//  Created by syahRiza on 11/24/16.
//  Copyright © 2016 Kii. All rights reserved.
//

#import <Foundation/Foundation.h>
@import KiiSDK;

@interface Snippet : NSObject
{
  NSString* groupUri;

}

@end
@implementation Snippet

-(void) snippet1_blocking{
  NSError *error = nil;

  // Instantiate the group.
  // (Assume that groupUri has the reference URI of the target group).
  KiiGroup *group = [KiiGroup groupWithURI:groupUri];
  [group refreshSynchronous:&error];
  if (error != nil) {
    // Error handling
    return;
  }

  NSString *topicname = @"GroupTopic";
  KiiTopic *topic = [group topicWithName:topicname];

  // Create APNs message fields
  KiiAPNSFields *apnsFields = [KiiAPNSFields createFields];

  // This sample code assumes that you are using the silent push notification,
  // so the AlertBody is not set.
  //[apnsFields setAlertBody:@"Show message"];

  // Enable the silent push notification by setting "content-available"
  [apnsFields setContentAvailable:@1];

  // Define category
  [apnsFields setCategory:@"MESSAGE_CATEGORY"];

  // Build a push message.
  // FCM and MQTT fields are set, so the message will be sent to devices using FCM or MQTT.
  KiiPushMessage *message = [KiiPushMessage composeMessageWithAPNSFields:apnsFields
                                                            andGCMFields:[KiiGCMFields createFields]
                                                           andMQTTFields:[KiiMQTTFields createFields]];

  NSMutableDictionary *data = [NSMutableDictionary dictionary];
  data[@"str"]=@"str";
  data[@"int"]=[NSNumber numberWithInt:1];
  data[@"bool"]=@YES;
  data[@"double"]=[NSNumber numberWithDouble:1.12];
  message.data = data;

  // Send the message.
  [topic sendMessageSynchronous:message
                      withError:&error];
  
  if (error != nil) {
    // Error handling
    return;
  }
}

-(void) snippet1_non_blocking{
  // Instantiate the group.
  // (Assume that groupUri has the reference URI of the target group).
  KiiGroup *group = [KiiGroup groupWithURI:groupUri];
  [group refreshWithBlock:^(KiiGroup *group, NSError *error) {
    if (error != nil) {
      // Error handling
      return;
    }

    // Instantiate the group topic.
    NSString *topicname = @"GroupTopic";
    KiiTopic *topic = [group topicWithName:topicname];

    // Create APNs message fields
    KiiAPNSFields *apnsFields = [KiiAPNSFields createFields];

    // This sample code assumes that you are using the silent push notification,
    // so the AlertBody is not set.
    //[apnsFields setAlertBody:@"Show message"];

    // Enable the silent push notification by setting "content-available"
    [apnsFields setContentAvailable:@1];

    // Define category
    [apnsFields setCategory:@"MESSAGE_CATEGORY"];

    // Build a push message.
    // FCM and MQTT fields are set, so the message will be sent to devices using FCM or MQTT.
    KiiPushMessage *message = [KiiPushMessage composeMessageWithAPNSFields:apnsFields
                                                              andGCMFields:[KiiGCMFields createFields]
                                                             andMQTTFields:[KiiMQTTFields createFields]];

    NSMutableDictionary *data = [NSMutableDictionary dictionary];
    data[@"str"]=@"str";
    data[@"int"]=[NSNumber numberWithInt:1];
    data[@"bool"]=@YES;
    data[@"double"]=[NSNumber numberWithDouble:1.12];
    message.data = data;

    // Send the message.
    [topic sendMessage:message
             withBlock:^(KiiTopic *topic, NSError *error) {
               if (error != nil) {
                 // Error handling
                 return;
               }
             }];
  }];
}
/*
 By default, the push message will be sent to all the push notification networks. You can control message delivery by setting a flag upon construction of a KiiPushMessage instance like in the following example:
 */
-(void) snippet1a{

  KiiPushMessage *message;

  // Deliver the push message to Android devices with using FCM only
  KiiGCMFields* gcmFields = [KiiGCMFields createFields];
  [gcmFields setSpecificData:@{@"fcmOnly":@"fcmOnlyData"}];

  message = [KiiPushMessage composeMessageWithAPNSFields:nil andGCMFields:gcmFields andMQTTFields:nil];

  // Deliver the push message to iOS devices with using APNS only
  KiiAPNSFields* apnsFields = [KiiAPNSFields createFields];
  [apnsFields setSpecificData:@{@"apnsOnly":@"apnsOnlyKey"}];

  message = [KiiPushMessage composeMessageWithAPNSFields:apnsFields andGCMFields:nil andMQTTFields:nil];

  // Deliver the push message to Thing devices with using FCM only
  KiiMQTTFields* mqttFields = [KiiMQTTFields createFields];
  [mqttFields setData:@{@"mqttOnly":@"apnsOnlyKey"}];

  message = [KiiPushMessage composeMessageWithAPNSFields:nil andGCMFields:nil andMQTTFields:mqttFields];
}

-(void) snippet2_blocking{
  NSError *error = nil;

  // Instantiate a user-scope topic.
  KiiUser* user = [KiiUser currentUser];
  NSString *topicname = @"MyTODO";
  KiiTopic *topic = [user topicWithName:topicname];

  // Create APNs message fields
  KiiAPNSFields *apnsFields = [KiiAPNSFields createFields];

  // This sample code assumes that you are using the silent push notification,
  // so the AlertBody is not set.
  //[apnsFields setAlertBody:@"Show message"];

  // Enable the silent push notification by setting "content-available"
  [apnsFields setContentAvailable:@1];

  // Define category
  [apnsFields setCategory:@"MESSAGE_CATEGORY"];

  // Build a push message.
  // FCM and MQTT fields are set, so the message will be sent to devices using FCM or MQTT.
  KiiPushMessage *message = [KiiPushMessage composeMessageWithAPNSFields:apnsFields
                                                            andGCMFields:[KiiGCMFields createFields]
                                                           andMQTTFields:[KiiMQTTFields createFields]];

  NSMutableDictionary *data = [NSMutableDictionary dictionary];
  data[@"Item"]=@"Do Something";
  data[@"Done"]=[NSNumber numberWithInt:1];
  message.data = data;

  // Send the push message.
  [topic sendMessageSynchronous:message
                      withError:&error];
  if (error != nil) {
    // Error handling
    return;
  }
}
-(void) snippet2_non_blocking{
  // Instantiate a user-scope topic.
  KiiUser* user = [KiiUser currentUser];
  NSString *topicname = @"MyTODO";
  KiiTopic *topic = [user topicWithName:topicname];

  // Create APNs message fields
  KiiAPNSFields *apnsFields = [KiiAPNSFields createFields];

  // This sample code assumes that you are using the silent push notification,
  // so the AlertBody is not set.
  //[apnsFields setAlertBody:@"Show message"];

  // Enable the silent push notification by setting "content-available"
  [apnsFields setContentAvailable:@1];

  // Define category
  [apnsFields setCategory:@"MESSAGE_CATEGORY"];

  // Build a push message.
  // FCM and MQTT fields are set, so the message will be sent to devices using FCM or MQTT.
  KiiPushMessage *message = [KiiPushMessage composeMessageWithAPNSFields:apnsFields
                                                            andGCMFields:[KiiGCMFields createFields]
                                                           andMQTTFields:[KiiMQTTFields createFields]];

  NSMutableDictionary *data = [NSMutableDictionary dictionary];
  data[@"Item"]=@"Do Something";
  data[@"Done"]=[NSNumber numberWithInt:1];
  message.data = data;

  // Send the push message.
  [topic sendMessage:message
           withBlock:^(KiiTopic *topic, NSError *error) {
             if (error != nil) {
               // Error handling
               return;
             }
           }];
}

/*
 For a user-scope topic, only the user who owns the topic can send a push message to it.

 Let's explore how to do this in the following sample code.
 */
-(void) snippet3_blocking{
  /*skip*/KiiTopic* topic;
  NSError *error = nil;

  // Create APNs message fields
  KiiAPNSFields *apnsFields = [KiiAPNSFields createFields];
  [apnsFields setAlertBody:@"short msg only!"];

  // Create a message.
  NSDictionary* commonData = @{};
  KiiPushMessage *pushMessage = [KiiPushMessage composeMessageWithAPNSFields:apnsFields
                                                                andGCMFields:nil
                                                               andMQTTFields:nil];
  pushMessage.data = commonData;
  
  // Send the push message.
  [topic sendMessageSynchronous:pushMessage
                      withError:&error];
  if (error != nil) {
    // Error handling
    return;
  }
}

//Saving the APNs payload
-(void) snippet3_non_blocking{
  /*skip*/KiiTopic* topic;
  // Create APNs message fields
  KiiAPNSFields *apnsFields = [KiiAPNSFields createFields];
  [apnsFields setAlertBody:@"short msg only!"];

  // Create a message.
  NSDictionary* commonData = @{};
  KiiPushMessage *pushMessage = [KiiPushMessage composeMessageWithAPNSFields:apnsFields
                                                                andGCMFields:nil
                                                               andMQTTFields:nil];
  pushMessage.data = commonData;

  // Send the push message.
  [topic sendMessage:pushMessage
           withBlock:^(KiiTopic *topic, NSError *error) {
             if (error != nil) {
               // Error handling
               return;
             }
           }];
}
/*
-(void) snippet2_blocking{
}
-(void) snippet2_non_blocking{
}
 */
@end

File: Snippet3/Snippet3/guides_ios_managing-push-notification_push-to-user_sending-messages.swift
//
//  ios_managing-push-notification_push-to-user_sending-messages.swift
//  Snippet
//
//  Created by syahRiza on 12/16/15.
//  Copyright © 2015 Kii. All rights reserved.
//

import Foundation
//MARK: path managing-push-notification/push-to-user/sending-messages/
private let groupUri = ""
//Sending Messages
private func snippet_1_blocking(){
  
  // Instantiate the group.
  // (Assume that groupUri has the reference URI of the target group).
  let group = KiiGroup(uri: groupUri)
  do{
    try group.refreshSynchronous()
  } catch let error as NSError {
    print(error.description)
    // Error handling
    return
  }
  
  // Instantiate the group topic.
  let topicName = "GroupTopic"
  let topic = group.topic(withName: topicName)
  
  // Create APNs message fields
  let apnsField = KiiAPNSFields.create()
  
  // This snippet assumes that you are using the silent push notification,
  // so the AlertBody is not set.
  // apnsField.alertBody = "Show Message"
  
  // Build a push message.
  // GCM fields are set nil, so the message will not send to Android devices.
  var dictionary = [AnyHashable: Any]()
  dictionary["from"] = "Alice"
  dictionary["msgBody"] = "Hi All"
  dictionary["Priority"] = NSNumber(value: 1 as Int32)
  dictionary["Urgent"] = NSNumber(value: false as Bool)
  dictionary["Weight"] = NSNumber(value: 1.12 as Double)
  apnsField.setSpecificData(dictionary)
  
  // Enable the silent push notification by setting "content-available"
  apnsField.contentAvailable = 1
  // Define category (iOS 8)
  apnsField.category = "MESSAGE_CATEGORY"
  
  let message = KiiPushMessage.composeMessage(with: apnsField, andGCMFields: nil)
  
  // Send the message.
  do{
    try topic.sendMessageSynchronous(message)
  } catch let error as NSError {
    print(error.description)
    // Error handling
    return
  }

}
private func snippet_1_non_blocking(){
  // Instantiate the group.
  // (Assume that groupUri has the reference URI of the target group).
  let group = KiiGroup(uri: groupUri)
  group.refresh { (group : KiiGroup?, error : Error?) -> Void in
    
    if error != nil {
      // Error handling
      return
    }
    // Instantiate the group topic.
    let topicName = "GroupTopic"
    let topic = group!.topic(withName: topicName)
    
    // Create APNs message fields
    let apnsField = KiiAPNSFields.create()
    
    // This snippet assumes that you are using the silent push notification,
    // so the AlertBody is not set.
    // apnsField.alertBody = "Show Message"
    
    // Build a push message.
    // GCM fields are set nil, so the message will not send to Android devices.
    var dictionary = [AnyHashable: Any]()
    dictionary["from"] = "Alice"
    dictionary["msgBody"] = "Hi All"
    dictionary["Priority"] = NSNumber(value: 1 as Int32)
    dictionary["Urgent"] = NSNumber(value: false as Bool)
    dictionary["Weight"] = NSNumber(value: 1.12 as Double)
    apnsField.setSpecificData(dictionary)
    
    // Enable the silent push notification by setting "content-available"
    apnsField.contentAvailable = 1
    // Define category (iOS 8)
    apnsField.category = "MESSAGE_CATEGORY"
    
    let message = KiiPushMessage.composeMessage(with: apnsField, andGCMFields: nil)
    
    // Send the message.
    topic.send(message, with: { (topic , error : Error?) -> Void in
      if error != nil {
        // Error handling
        return
      }
    })
  }
  
}
//Let's explore how to do this in the following sample snippet.
private func snippet_2_blocking(){
  
  
  // Instantiate a user-scope topic.
  let user = KiiUser.current()!
  let topicName = "MyToDO"
  let topic = user.topic(withName: topicName)
  
  // Create APNs message fields
  let apnsField = KiiAPNSFields.create()
  
  // This snippet assumes that you are using the silent push notification,
  // so the AlertBody is not set.
  // apnsField.alertBody = "Show Message"
  
  // Build a push message.
  // GCM fields are set nil, so the message will not send to Android devices.
  var dictionary = [AnyHashable: Any]()
  dictionary["item"] = "Do Something"
  dictionary["Done"] = NSNumber(value: 1 as Int32)
  
  apnsField.setSpecificData(dictionary)
  
  // Enable the silent push notification by setting "content-available"
  apnsField.contentAvailable = 1
  // Define category (iOS 8)
  apnsField.category = "MESSAGE_CATEGORY"
  
  let message = KiiPushMessage.composeMessage(with: apnsField, andGCMFields: nil)
  
  // Send the message.
  do{
    try topic.sendMessageSynchronous(message)
  } catch let error as NSError {
    print(error.description)
    // Error handling
    return
  }

}
private func snippet_2_non_blocking(){
  // Instantiate a user-scope topic.
  let user = KiiUser.current()!
  let topicName = "MyToDO"
  let topic = user.topic(withName: topicName)
  
  // Create APNs message fields
  let apnsField = KiiAPNSFields.create()
  
  // This snippet assumes that you are using the silent push notification,
  // so the AlertBody is not set.
  // apnsField.alertBody = "Show Message"
  
  // Build a push message.
  // GCM fields are set nil, so the message will not send to Android devices.
  var dictionary = [AnyHashable: Any]()
  dictionary["item"] = "Do Something"
  dictionary["Done"] = NSNumber(value: 1 as Int32)
  apnsField.setSpecificData(dictionary)
  
  // Enable the silent push notification by setting "content-available"
  apnsField.contentAvailable = 1
  // Define category (iOS 8)
  apnsField.category = "MESSAGE_CATEGORY"
  
  let message = KiiPushMessage.composeMessage(with: apnsField, andGCMFields: nil)
  
  // Send the message.
  topic.send(message, with: { (topic , error : Error?) -> Void in
    if error != nil {
      // Error handling
      return
    }
  })
}
private let topic = KiiTopic()
//Saving the APNS payload
//short
private func snippet_3_blocking(){
  
  
  // Create APNs message fields
  let apnsField = KiiAPNSFields.create()
  apnsField.alertBody = "short message only"
  
  let message = KiiPushMessage.composeMessage(with: apnsField, andGCMFields: nil)
  
  // Send the message.
  do{
    try topic.sendMessageSynchronous(message)
  } catch let error as NSError {
    print(error.description)
    // Error handling
    return
  }
  
}
private func snippet_3_non_blocking(){
  // Create APNs message fields
  let apnsField = KiiAPNSFields.create()
  apnsField.alertBody = "short message only"
  
  // Create a message.
  let message = KiiPushMessage.composeMessage(with: apnsField, andGCMFields: nil)
  
  // Send the message.
  topic.send(message, with: { (topic , error : Error?) -> Void in
    if error != nil {
      // Error handling
      return
    }
  })
  
}

//short
private func snippet_4_blocking(){
  
  
  // Create APNs message fields
  let apnsField = KiiAPNSFields.create()
  apnsField.alertBody = "Looooooooooooooooooong Message!!"
  
  // Create a message.
  let message = KiiPushMessage.composeMessage(with: apnsField, andGCMFields: nil)
  //disable "s" field
  message.sendSender = NSNumber(value: false as Bool)
  //disable "w" field
  message.sendWhen = NSNumber(value: false as Bool)
  //disable "to" field
  message.sendTopicID = NSNumber(value: false as Bool)
  // Disable "sa", "st" and "su" field
  message.sendObjectScope = NSNumber(value: false as Bool)
  
  // Send the message.
  do{
    try topic.sendMessageSynchronous(message)
  } catch let error as NSError {
    print(error.description)
    // Error handling
    return
  }

}
private func snippet_4_non_blocking(){
  // Create APNs message fields
  let apnsField = KiiAPNSFields.create()
  apnsField.alertBody = "Looooooooooooooooooong Message!!"
  
  // Create a message.
  let message = KiiPushMessage.composeMessage(with: apnsField, andGCMFields: nil)
  
  //disable "s" field
  message.sendSender = NSNumber(value: false as Bool)
  //disable "w" field
  message.sendWhen = NSNumber(value: false as Bool)
  //disable "to" field
  message.sendTopicID = NSNumber(value: false as Bool)
  // Disable "sa", "st" and "su" field
  message.sendObjectScope = NSNumber(value: false as Bool)
  
  // Send the message.
  topic.send(message, with: { (topic , error : Error?) -> Void in
    if error != nil {
      // Error handling
      return
    }
  })
}
